src/lib/Optimize.hs

```diff
@@ -61,3 +68,165 @@ instance GenericTraverser UTLS where
+
+-- The interesting instances
+
+instance Color c => HasDCE (Name c) where
+  dce n = modify (<> FV (freeVarsE n)) $> n
+
+instance HasDCE Block where
+  dce (Block ann decls ans) = case (ann, decls) of
+    (NoBlockAnn      , Empty) -> Block NoBlockAnn Empty <$> dce ans
+    (NoBlockAnn      , _    ) -> error "should be unreachable"
+    (BlockAnn ty effs, _    ) -> do
+      -- The free vars accumulated in the state of DCEM should correspond to
+      -- the free vars of the Abs of the block answer, by the decls traversed
+      -- so far. dceNest takes care to uphold this invariant, but we temporarily
+      -- reset the state to an empty map, just so that names from the surrounding
+      -- block don't end up influencing elimination decisions here. Note that we
+      -- restore the state (and accumulate free vars of the DCE'd block into it)
+      -- right after dceNest.
+      old <- get
+      put mempty
+      Abs decls' ans' <- dceNest decls ans
+      modify (<> old)
+      ty' <- dce ty
+      effs' <- dce effs
+      return $ Block (BlockAnn ty' effs') decls' ans'
+
+data CachedFVs e n = UnsafeCachedFVs { _cachedFVs :: (NameSet n), fromCachedFVs :: (e n) }
+instance HoistableE (CachedFVs e) where
+  freeVarsE (UnsafeCachedFVs fvs _) = fvs
+
+wrapWithCachedFVs :: HoistableE e => e n -> DCEM n (CachedFVs e n)
+wrapWithCachedFVs e = do
+  FV fvs <- get
+#ifdef DEX_DEBUG
+  let fvsAreCorrect = nameSetRawNames fvs == nameSetRawNames (freeVarsE e)
+#else
+  -- Verification of this invariant defeats the performance benefits of
+  -- avoiding the extra traversal (e.g. actually having linear complexity),
+  -- so we only do that in debug builds.
+  let fvsAreCorrect = True
+#endif
+  case fvsAreCorrect of
+    True -> return $ UnsafeCachedFVs fvs e
+    False -> error "Free variables were computed incorrectly."
+
+hoistUsingCachedFVs :: (BindsNames b, HoistableE e) => b n l -> e l -> DCEM l (HoistExcept (e n))
+hoistUsingCachedFVs b e = do
+  ec <- wrapWithCachedFVs e
+  return $ case hoist b ec of
+    HoistSuccess e' -> HoistSuccess $ fromCachedFVs e'
+    HoistFailure err -> HoistFailure err
+
+data ElimResult n where
+  ElimSuccess :: Abs (Nest Decl) Atom n -> ElimResult n
+  ElimFailure :: Decl n l -> Abs (Nest Decl) Atom l -> ElimResult n
+
+dceNest :: Nest Decl n l -> Atom l -> DCEM n (Abs (Nest Decl) Atom n)
+dceNest decls ans = case decls of
+  Empty -> Abs Empty <$> dce ans
+  Nest b@(Let _ decl) bs -> do
+    isPureDecl <- isPure decl
+    dceAttempt <- refreshAbs (Abs b (Abs bs ans)) \b' (Abs bs' ans') -> do
+      below <- dceNest bs' ans'
+      case isPureDecl of
+        False -> return $ ElimFailure b' below
+        True  -> do
+          hoistUsingCachedFVs b' below <&> \case
+            HoistSuccess below' -> ElimSuccess below'
+            HoistFailure _ -> ElimFailure b' below
+    case dceAttempt of
+      ElimSuccess below' -> return below'
+      ElimFailure (Let b' decl') (Abs bs'' ans'') -> do
```

Comment:
`b'` is only in scope in the `Abs` that appears here on the right, and the free vars are removed when we exit the `refreshAbs` above. `decl'` cannot refer to `b'`, so even though we traverse it, it will not add _this `b'`_ into the free vars. But I just realized I forgot to add the free vars of `b'` itself (which was ok, because a `NameBinder` never has any free vars).